### PR TITLE
QASM benchmarks

### PR DESCRIPTION
**Pull request by ecpeterson**
Contains a new routine (and `make` target) for benchmarking quilc's performance against the standard QASM benchmarking suite. Partially? addresses #465.

Adds the QASM benchmarks from the IBM QX Mapper as a git submodule.

I'm a little uncomfortable adding yet another benchmarking code path to quilc. This could probably be implemented within the framework set up in `rewiring-analysis.lisp`, but I also don't feel that that's much of an improvement over this minimalist procedure. On the other hand, one way that this implementation falls short is that it prints its results straight to `*standard-output*`, without opportunity to manipulate the results as data. I'd support at least opening an issue to smooth out and standardize the benchmarking techniques within quilc.

Example output:
```
+------------------+----------+-------+----------+
|       NAME       | TIME (s) | SWAPS | 2Q DEPTH |
+------------------+----------+-------+----------+
| 0410184_169      |    4.292 |   115 |      185 |
| 3_17_13          |    0.778 |    11 |       35 |
| 4_49_16          |    3.578 |    84 |      205 |
| 4gt10-v1_81      |    2.775 |    53 |      128 |
...
| z4_268           | >>>>>>30 | ????? | ???????? |
+------------------+----------+-------+----------+
```

### CODE REVIEW

File: benchmarking/qasm-benchmarks.lisp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+;;;;
+;;;; Author: Eric Peterson
+
+(in-package #:cl-quil-benchmarking)
+
+(defvar *qasm-benchmark-path*
+  (asdf:system-relative-pathname :cl-quil "benchmarking/ibm_qx_mapping/examples/"))
+
+(defun qasm-test-files ()
+  (uiop:directory-files *qasm-benchmark-path* #P"*.qasm"))
```
Comment (stylewarning) [inline benchmarking/qasm-benchmarks.lisp]: Could we check that the directory exists in this function, and if not, error with a handsome message?

File: benchmarking/qasm-benchmarks.lisp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            :do (quil::append-instruction-to-lschedule lschedule instr)
+          :finally (return (quil::lscheduler-calculate-depth lschedule)))))
+
+(defun benchmark-qasm-suite (&key (timeout 30))
+  (format t "+------------------+----------+-------+----------+~%")
+  (format t "|       NAME       | TIME (s) | SWAPS | 2Q DEPTH |~%")
+  (format t "+------------------+----------+-------+----------+~%")
+  (let ((chip (quil::build-ibm-qx5))
+        (quil::*default-addresser-state-class* 'quil::temporal-addresser-state))
+    (dolist (file (qasm-test-files))
```
Comment (stylewarning) [inline benchmarking/qasm-benchmarks.lisp]: `(finish-output)` after each dolist iteration

File: benchmarking/qasm-benchmarks.lisp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                  (format t "| ~Vf | ~Vd | ~Vd |~%"
+                          8 (/ elapsed-time 1000)
+                          5 swaps
+                          8 (calculate-multiqubit-gate-depth (parsed-program-executable-code cpp)))))))
+        (bt:timeout ()
+          (format t "| ~8,'>d | ????? | ???????? |~%"
+                  timeout)
+          )
+        )))
+  (format t "+------------------+----------+-------+----------+~%"))
```
Comment (stylewarning) [inline benchmarking/qasm-benchmarks.lisp]: finish-output after this for good health

File: benchmarking/qasm-benchmarks.lisp [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                                                      :originating-file (first (directory file)))
+                                         chip
+                                         :protoquil t)
+                  (format t "| ~Vf | ~Vd | ~Vd |~%"
+                          8 (/ elapsed-time 1000)
+                          5 swaps
+                          8 (calculate-multiqubit-gate-depth (parsed-program-executable-code cpp)))))))
+        (bt:timeout ()
+          (format t "| ~8,'>d | ????? | ???????? |~%"
+                  timeout)
```
Comment (stylewarning) [inline benchmarking/qasm-benchmarks.lisp]: close parens

File: benchmarking/qasm-benchmarks.lisp [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                    (quil::compiler-hook (quil::parse text
+                                                      :originating-file (first (directory file)))
+                                         chip
+                                         :protoquil t)
+                  (format t "| ~Vf | ~Vd | ~Vd |~%"
+                          8 (/ elapsed-time 1000)
+                          5 swaps
+                          8 (calculate-multiqubit-gate-depth (parsed-program-executable-code cpp)))))))
+        (bt:timeout ()
+          (format t "| ~8,'>d | ????? | ???????? |~%"
```
Comment (notmgsk) [inline benchmarking/qasm-benchmarks.lisp]: `?????`?

Comment (stylewarning) [reply benchmarking/qasm-benchmarks.lisp]: if the benchmark times out, ur left with ❓❓❓❓❓

**Comment by stylewarning** [comment]
One other thing: It might be nice to have a function `describe-quilc-configuration` which describes the set of options, parameters, etc. that quilc is configured to use, and that can be printed out before the b-marks are run

**Comment by ecpeterson** [comment]
> One other thing: It might be nice to have a function `describe-quilc-configuration` which describes the set of options, parameters, etc. that quilc is configured to use, and that can be printed out before the b-marks are run

Is there a way to get a list of parameter definitions within a package, rather than maintaining a list of all possible config variables?

File: benchmarking/qasm-benchmarks.lisp (hunk 4 above)
Comment (ecpeterson) [reply benchmarking/qasm-benchmarks.lisp]: oh, weird, ty

File: benchmarking/qasm-benchmarks.lisp (hunk 3 above)
Comment (ecpeterson) [reply benchmarking/qasm-benchmarks.lisp]: as u please

File: benchmarking/qasm-benchmarks.lisp (hunk 5 above)
Comment (ecpeterson) [reply benchmarking/qasm-benchmarks.lisp]: I'm happy to replace this with `DNF` or something.

**Comment by stylewarning** [comment]
> > One other thing: It might be nice to have a function `describe-quilc-configuration` which describes the set of options, parameters, etc. that quilc is configured to use, and that can be printed out before the b-marks are run
> 
> Is there a way to get a list of parameter definitions within a package, rather than maintaining a list of all possible config variables?

here's a lil somethin somethin for u

```
(defun idiomatic-package-variables (package &key include-internal)
  (let ((symbols nil))
    (labels ((bookended-by (char symbol)
               (let* ((string (symbol-name symbol))
                      (len (length string)))
                 (and (<= 2 len)
                      (char= char (char string 0) (char string (1- len))))))
             (variable-looking-symbol-p (s)
               (and (boundp s)
                    (or (bookended-by #\+ s)
                        (bookended-by #\* s))))
             (ponder-symbol (s)
               (when (variable-looking-symbol-p s)
                 (push s symbols))))
      (do-symbols (s package (sort symbols #'string<))
        (multiple-value-bind (s status)
            (find-symbol (symbol-name s) package)
          (case status
            (:internal
             (when include-internal
               (ponder-symbol s)))
            (:external
             (ponder-symbol s))))))))
```

Examples:

```
CL-USER> (idiomatic-package-variables ':qvm)
(QVM:*COMPILE-BEFORE-RUNNING* QVM:*QUBITS-REQUIRED-FOR-PARALLELIZATION*
 QVM:*TRANSITION-VERBOSE*)

CL-USER> (idiomatic-package-variables ':qvm :include-internal t)
(QVM::**ALIGNMENT** QVM::**APPLY-MATRIX-OPERATOR-FUNCTIONS**
 QVM::**CLASSICAL-MEMORY-SIZE-LIMIT** QVM::**CLIFFORD-OPERATIONS**
 QVM::**CLIFFORDS** QVM::**DOTIMES-ITERATOR** QVM::**INTEGER-BITS**
 QVM::**REAL-BITS** QVM::**SHARED-MEMORIES** QVM::*AVAILABLE-KERNELS*
 QVM:*COMPILE-BEFORE-RUNNING* QVM::*COMPILE-MEASURE-CHAINS*
 QVM::*FUSE-GATES-DURING-COMPILATION*
 QVM::*INLINE-STATIC-GATES-DURING-COMPILATION* QVM::*OPTIMIZE-BRISKLY*
 QVM::*OPTIMIZE-DANGEROUSLY-FAST* QVM::*OPTIMIZE-SAFELY*
 QVM::*QUBIT-LIMIT-FOR-USING-SERIAL-KERNELS*
 QVM:*QUBITS-REQUIRED-FOR-PARALLELIZATION* QVM:*TRANSITION-VERBOSE*
 QVM::+DEFAULT-NOISE-PREDICATE-NAME+ QVM::+MAX-NAT-TUPLE-CARDINALITY+
 QVM::+MAXPRIORITY+ QVM::+MINPRIORITY+ QVM::+OCTETS-PER-CFLONUM+
 QVM::+OCTETS-PER-FLONUM+)
```

File: benchmarking/qasm-benchmarks.lisp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        (format t "| ~Va " 16 (trim-long-string (pathname-name file) 16))
+        (finish-output)
+        (handler-case
+            (bordeaux-threads:with-timeout (timeout)
+              (let ((text (alexandria:read-file-into-string file)))
+                #+sbcl (sb-ext:gc :full t)
+                (with-stopwatch elapsed-time
+                  (multiple-value-bind (cpp swaps)
+                      (quil::compiler-hook (quil::parse text
+                                                        :originating-file (first (directory file)))
```
Comment (appleby) [inline benchmarking/qasm-benchmarks.lisp]: what's up with this `(first (directory file))`?

File: qelib1.inc [hunk 7]
```diff
@@ -0,0 +1,95 @@
+// Quantum Experience (QE) Standard Header
```
Comment (appleby) [inline qelib1.inc]: There is another copy of this file in tests/qasm-files/qelib1.inc

File: benchmarking/qasm-benchmarks.lisp (hunk 6 above)
Comment (ecpeterson) [reply benchmarking/qasm-benchmarks.lisp]: Copy/pasted out of another benchmarking file where it also doesn't make sense.

File: qelib1.inc (hunk 7 above)
Comment (ecpeterson) [reply qelib1.inc]: Figured out how to redirect includes to use that one instead.

File: benchmarking/qasm-benchmarks.lisp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+            (:internal
+             (when include-internal
+               (ponder-symbol s)))
+            (:external
+             (ponder-symbol s))))))))
+
+(defun print-configuration ()
+  (format t "Current CL-QUIL configuration:~%")
+  (dolist (symb (idiomatic-package-variables ':cl-quil :include-internal t))
+    (format t "~a: ~a~%" symb (eval symb)))
```
Comment (stylewarning) [inline benchmarking/qasm-benchmarks.lisp]:
```suggestion
    (format t "~a: ~a~%" symb (symbol-value symb)))
```

**Pull request merged**